feat(linter): allow adding more overrides via `extends` configs

camchenry opened the pull request:
- closes https://github.com/oxc-project/oxc/issues/9307

Implements basic support for appending to the `overrides` array via configuration files referenced with `extends`.

camchenry commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9475?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9475** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9475?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9475?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9473** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9473?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9472** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9472?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9308** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9308?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9217** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9217?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9156** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9156?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9155** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9155?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9153** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9153?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9152** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9152?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9475).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Sysix commented on crates/oxc_linter/src/config/config_builder.rs:
```diff
@@ -134,6 +137,13 @@ impl ConfigStoreBuilder {
+                                let overrides =
+                                    std::mem::take(&mut extended_config_store.overrides);
+                                for r#override in overrides {
+                                    builder = builder.with_override(r#override);
```
Can we add a test for `extends` and `files` combination which targets the same directory:
Example:

`.oxlintrc.json`
````json
{
  "rules": { "no-debugger": "error"  },
  "overrides": [
    {
      "files": ["config/*.ts", "config/*.js"],
      "rules": { "no-debugger": "off" }
    }
  ]
}
````

and `config/.oxclintrc.json`:

````json
{
  "extends": ["../.oxlintrc.json"],
  "rules": { "no-debugger": "warn" },
}
````

camchenry replied:
Thanks, I've added this test case. I checked too and our behavior here is aligned with eslint too (this results in warn on files in `config`)

camchenry committed:
feat(linter): allow adding more overrides via `extends` configs (#9475)

- closes https://github.com/oxc-project/oxc/issues/9307

Implements basic support for appending to the `overrides` array via configuration files referenced with `extends`.
